tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java

```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
     private void insertData(TableName tableName, JdbcTableDataSource dataSource)
     {
         Iterator<List<Object>> dataRows = dataSource.getDataRows();
-        PreparedStatement preparedStatement = null;
-        List<List<Object>> batchRows = new ArrayList<>();
+        if (!dataRows.hasNext()) {
+            return;
+        }
+        try {
+            insertData(tableName, dataRows);
+        }
+        catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void insertData(TableName tableName, Iterator<List<Object>> dataRows)
+            throws SQLException
+    {
+        Loader loader = null;
         try {
-            int rowsInserted = 0;
-            while (dataRows.hasNext()) {
-                List<Object> values = dataRows.next();
-                if (preparedStatement == null) {
-                    preparedStatement = buildPreparedStatementForInsert(tableName.getNameInDatabase(), values.size());
+            for (List<List<Object>> batch : asBatches(dataRows)) {
+                if (loader == null) {
+                    int columnsCount = batch.get(0).size();
+                    loader = new LoaderFactory().create(queryExecutor, tableName.getNameInDatabase(), columnsCount);
                 }
+                loader.load(batch);
+            }
+        }
+        finally {
+            if (loader != null) {
+                loader.close();
+            }
+        }
+    }
+
+    private static Iterable<List<List<Object>>> asBatches(Iterator<List<Object>> dataRows)
+    {
+        return () -> new Iterator<List<List<Object>>>()
+        {
+            @Override
+            public boolean hasNext()
+            {
+                return dataRows.hasNext();
+            }
+
+            @Override
+            public List<List<Object>> next()
+            {
+                List<List<Object>> batch = new ArrayList<>();
+                while (dataRows.hasNext() || batch.size() < BATCH_SIZE) {
```

Comment:
If `dataRows` is empty (`hasNext() = false`) but `batch.size()` < `BATCH_SIZE`, there will be `NoSuchElementException` thrown in next line.